File: apps/windows/TeslaSync.App/shared-surfaces/Speed.Model.cs
using System.Globalization;
using TeslaSync.App.Core.Units;

namespace TeslaSync.App.SharedSurfaces;

/// <summary>
/// Canonical metadata + conversion constants for the Speed surface — the native analogue of the
/// module-level constants in <c>web/src/components/data-display/format/Speed.tsx</c>. The web component is a
/// pure presentational speed readout: it accepts a caller-supplied speed in <c>mph</c> or <c>kmh</c>, folds it
/// to SI metres-per-second, reconverts to the user's <see cref="UnitPref.Speed"/> via <c>convertSpeedFromSI</c>,
/// and renders <c>{fmtNumber(value, precision)} {speedUnit}</c> with a hover title showing the raw
/// caller-supplied value in its source unit. It reads no network data and renders no titles/labels of its own,
/// so this carries only the diagnostics slug, the automation id, the SI fold factors the source hard-codes, and
/// the formatting defaults (<c>fmtNumber</c>'s global precision and the <c>toFixed(1)</c> source-title precision).
/// </summary>
public static class SpeedRegistration
{
    /// <summary>Diagnostics surface slug emitted with the <c>view.opened</c> event (P1/S11 contract).</summary>
    public const string Slug = "Speed";

    /// <summary>
    /// The root automation id the view stamps on itself. The web component declares no <c>data-testid</c>
    /// (it is an anonymous inline span), so this is the native-only stable hook UI-automation tests target.
    /// </summary>
    public const string RootAutomationId = "speed";

    /// <summary>
    /// Default fraction digits when the caller supplies no <c>precision</c>. Mirrors the web
    /// <c>fmtNumber(value, precision)</c> fallback to the module-level <c>_globalPrecision</c>, whose default is
    /// <c>2</c> and which <c>useSettings</c> seeds from <c>settings.decimal_precision ?? 2</c>
    /// (web/src/lib/numberFormat.ts, web/src/hooks/useSettings.ts).
    /// </summary>
    public const int DefaultPrecision = 2;

    /// <summary>Fraction digits used by the hover title (web <c>mph.toFixed(1)</c> / <c>kmh.toFixed(1)</c>).</summary>
    public const int SourceTitlePrecision = 1;

    /// <summary>Metres-per-second in one mile-per-hour (web <c>mph * 0.44704</c>; 1609.344 m / 3600 s, NIST).</summary>
    public const double MetersPerSecondPerMph = 0.44704;

    /// <summary>Metres in one kilometre (web folds km/h via <c>(kmh * 1000) / 3600</c>).</summary>
    public const double MetersPerKilometer = 1000.0;

    /// <summary>Seconds in one hour (web folds km/h via <c>(kmh * 1000) / 3600</c>).</summary>
    public const double SecondsPerHour = 3600.0;
}

/// <summary>
/// Identifies which caller-supplied input produced a <see cref="SpeedProjection"/> — the native analogue of the
/// web component's <c>mph</c>-first / <c>kmh</c>-fallback / neither branch
/// (web/src/components/data-display/format/Speed.tsx).
/// </summary>
public enum SpeedSource
{
    /// <summary>No finite input was supplied; the readout shows the empty fallback and has no title.</summary>
    None = 0,

    /// <summary>The <c>mph</c> input was used (web <c>mph != null &amp;&amp; Number.isFinite(mph)</c>).</summary>
    MilesPerHour,

    /// <summary>The <c>kmh</c> input was used (web <c>else if (kmh != null &amp;&amp; Number.isFinite(kmh))</c>).</summary>
    KilometersPerHour,
}

/// <summary>
/// Pure projection of the readout's render inputs — the native port of the web component body
/// (web/src/components/data-display/format/Speed.tsx). It picks the source input (mph first, then km/h), folds it
/// to SI metres-per-second using the exact factors the web hard-codes, reconverts to the user's display unit via
/// <see cref="UnitConverters.SpeedFromSi"/> (the <c>convertSpeedFromSI</c> port), and formats the number through
/// <see cref="ScalarFormatters.FormatNumber(double?, int, string)"/> — the verified 1:1 port of the web
/// <c>fmtNumber(value, precision)</c> (en-US grouping, fixed fraction digits). The hover <see cref="Title"/> mirrors
/// the web <c>title</c> attribute: the raw caller-supplied value rendered with <c>toFixed(1)</c> plus its source
/// unit. When neither input is finite the projection is the empty fallback (web bare <c>—</c> span, no title).
/// Kept a side-effect-free value type so the adapter is unit-testable without a view-model or a UI thread; the
/// <see cref="SpeedViewModel"/> and the WinUI view both render from it.
/// </summary>
public readonly record struct SpeedProjection
{
    private SpeedProjection(
        SpeedSource source,
        bool hasValue,
        string displayText,
        string? title,
        SpeedUnit unit,
        int precision)
    {
        Source = source;
        HasValue = hasValue;
        DisplayText = displayText;
        Title = title;
        Unit = unit;
        Precision = precision;
    }

    /// <summary>Which caller input produced this projection (mph, km/h, or none).</summary>
    public SpeedSource Source { get; }

    /// <summary>
    /// Whether a finite speed was supplied. False renders the empty fallback (web <c>—</c> branch); true
    /// renders the converted value with its unit.
    /// </summary>
    public bool HasValue { get; }

    /// <summary>
    /// The full visible readout. <c>"{number} {unit}"</c> (e.g. <c>"60.00 mph"</c>) when <see cref="HasValue"/>,
    /// otherwise the empty fallback (em dash). This is the surface's accessible name.
    /// </summary>
    public string DisplayText { get; }

    /// <summary>
    /// The hover title — the raw caller-supplied value in its source unit (web <c>title</c>:
    /// <c>"{mph.toFixed(1)} mph"</c> / <c>"{kmh.toFixed(1)} km/h"</c>). <see langword="null"/> in the empty state,
    /// matching the web span that renders no title when there is no value.
    /// </summary>
    public string? Title { get; }

    /// <summary>The display unit the value was converted to (web <c>unitPrefs.speed</c>).</summary>
    public SpeedUnit Unit { get; }

    /// <summary>The fraction-digit count the number was rendered with (resolved precision; never negative).</summary>
    public int Precision { get; }

    /// <summary>
    /// The accessible name the view exposes to Narrator — the visible readout, so the screen reader announces the
    /// meaningful converted value (web inline span text). Equal to <see cref="DisplayText"/>.
    /// </summary>
    public string AccessibleName => DisplayText;

    /// <summary>
    /// Project the render inputs exactly as the web component body does. <paramref name="mph"/> is preferred when
    /// finite; otherwise <paramref name="kmh"/> when finite; otherwise the empty fallback. The supplied value is
    /// folded to SI metres-per-second, reconverted to <paramref name="pref"/>'s speed unit, and formatted with the
    /// resolved precision (<paramref name="precision"/> when supplied, else <see cref="UnitPref.Precision"/>, else
    /// <see cref="SpeedRegistration.DefaultPrecision"/> — the web <c>precision ?? _globalPrecision</c> chain).
    /// </summary>
    /// <param name="mph">The speed in miles-per-hour (web <c>mph</c>), or null.</param>
    /// <param name="kmh">The speed in kilometres-per-hour (web <c>kmh</c>), or null.</param>
    /// <param name="precision">The per-call fraction-digit override (web <c>precision</c>), or null.</param>
    /// <param name="pref">The user's unit preference bag (web <c>useUnits().unitPrefs</c>).</param>
    public static SpeedProjection Project(double? mph, double? kmh, int? precision, UnitPref pref)
    {
        ArgumentNullException.ThrowIfNull(pref);

        SpeedUnit unit = pref.Speed;

        double sourceValue;
        double metersPerSecond;
        SpeedSource source;
        if (IsFinite(mph))
        {
            sourceValue = mph!.Value;
            metersPerSecond = sourceValue * SpeedRegistration.MetersPerSecondPerMph;
            source = SpeedSource.MilesPerHour;
        }
        else if (IsFinite(kmh))
        {
            sourceValue = kmh!.Value;
            metersPerSecond = sourceValue * SpeedRegistration.MetersPerKilometer / SpeedRegistration.SecondsPerHour;
            source = SpeedSource.KilometersPerHour;
        }
        else
        {
            // web: `if (sourceMph == null) return <span className={className}>—</span>;` — em-dash fallback, no title.
            string empty = pref.EmptyDisplay ?? UnitFormatters.DefaultEmptyDisplay;
            return new SpeedProjection(SpeedSource.None, hasValue: false, empty, title: null, unit, precision: 0);
        }

        int resolved = ResolvePrecision(precision, pref);
        double display = UnitConverters.SpeedFromSi(metersPerSecond, unit);
        string number = ScalarFormatters.FormatNumber(display, resolved);
        string displayText = $"{number} {UnitLabels.Label(unit)}";
        string title = $"{FormatFixed(sourceValue, SpeedRegistration.SourceTitlePrecision)} {SourceLabel(source)}";
        return new SpeedProjection(source, hasValue: true, displayText, title, unit, resolved);
    }

    private static string SourceLabel(SpeedSource source) => source switch
    {
        // The title shows the caller's source unit verbatim (web template literals), not the display unit.
        SpeedSource.MilesPerHour => "mph",
        SpeedSource.KilometersPerHour => "km/h",
        _ => string.Empty,
    };

    private static int ResolvePrecision(int? over, UnitPref pref)
    {
        // web fmtNumber(value, precision): `const d = precision ?? _globalPrecision`. A supplied precision wins
        // (even 0); otherwise the user's configured precision (UnitPref.Precision — the _globalPrecision analog);
        // otherwise the documented default of 2. Negatives are clamped to 0 (ScalarFormatters does the same).
        if (over is { } o)
        {
            return o < 0 ? 0 : o;
        }

        if (pref.Precision is { } p)
        {
            return p < 0 ? 0 : p;
        }

        return SpeedRegistration.DefaultPrecision;
    }

    private static bool IsFinite(double? v) => v is { } d && !double.IsNaN(d) && !double.IsInfinity(d);

    private static string FormatFixed(double value, int digits)
    {
        // JS Number.prototype.toFixed(digits): fixed-point, '.' separator, NO thousands grouping, round half away
        // from zero. Reproduced with AwayFromZero rounding + invariant fixed-point formatting.
        double rounded = Math.Round(value, digits, MidpointRounding.AwayFromZero);
        return rounded.ToString("F" + digits.ToString(CultureInfo.InvariantCulture), CultureInfo.InvariantCulture);
    }
}

/// <summary>
/// PII-safe diagnostics for the Speed surface (P1/S11 diagnostics contract). The readout carries no user content
/// beyond a caller-supplied number, so the collector records only the operational <c>view.opened</c> event with
/// the surface slug — never the value or its unit. Thread-safe.
/// </summary>
public sealed class SpeedDiagnostics
{
    private readonly Action<string>? _sink;
    private long _viewsOpened;

    /// <summary>Creates the collector over an optional PII-safe diagnostics sink.</summary>
    /// <param name="sink">The sink the <c>view.opened</c> line is written to, or null.</param>
    public SpeedDiagnostics(Action<string>? sink = null) => _sink = sink;

    /// <summary>Number of times the surface has been opened.</summary>
    public long ViewsOpened => Interlocked.Read(ref _viewsOpened);

    /// <summary>Record that the surface was opened, emitting <c>view.opened slug=Speed</c>.</summary>
    public void RecordViewOpened()
    {
        Interlocked.Increment(ref _viewsOpened);
        _sink?.Invoke($"view.opened slug={SpeedRegistration.Slug}");
    }
}

File: apps/android/app/src/main/java/com/teslasync/shared-surfaces/Speed/SpeedModel.kt
// Pure, framework-free model + projection + diagnostics for the Speed shared surface — the native analogue of
// web/src/components/data-display/format/Speed.tsx. No Compose, no Android framework, no HTTP: every declaration
// here is exercised off-device in the :app:testReleaseUnitTest gate, keeping the composable a thin render layer
// over these pure functions (the accepted sibling-surface contract used by AnimatedNumber / SpeedGearPanel).
//
// The web source is a PURELY PRESENTATIONAL speed renderer, not a data-fetching view. It accepts a caller value in
// one of two source units (`mph` OR `kmh`) plus a `precision`, reads the user's display preference through
// `useUnits`, and renders the value converted to that preference with the unit symbol appended — or an em dash
// when no finite value is supplied. Crucially the web converts the caller input to SI metres-per-second FIRST
// (`mph * 0.44704` or `kmh * 1000 / 3600`; the misnamed `sourceMph` actually holds m/s), then converts SI → the
// display unit exactly once via `convertSpeedFromSI`, and formats with the web global `fmtNumber` (whose default
// precision is the user's `decimal_precision`, 2 when unset — NOT the lib formatter's 0). This file owns that
// math: the two source-unit → SI factors, the single SI → display conversion (the shared `convertSpeedFromSI`),
// the `fmtNumber`-parity formatter, the hover-title assembly (web `title` = the raw caller value with its source
// unit, `toFixed(1)`), and the em-dash empty branch.
//
// The web source has NO async cache-then-network feed — it is handed a finished number — so there is no loading /
// error / stale / offline lifecycle to model; inventing one would fabricate behaviour the web spec does not have,
// exactly as the accepted AnimatedNumber / VisuallyHidden presentational ports document. The surface's real,
// reproduced states are: a value in the imperial preference, a value in the metric preference, the km/h source
// path, the precision / unit format variants, and the no-value em dash. The web renders no static copy of its own
// — the number, the display unit symbol (derived from the preference), and the source unit symbol in the title
// are its only text — so the surface carries NO i18n keys; there is none to map, and none is invented.
//
// `InvalidPackageDeclaration` is suppressed because this surface's mandated directory
// (com/teslasync/shared-surfaces/Speed — the P3 prompt's allowed-files path) cannot form a valid Kotlin package
// (a hyphen segment and a PascalCase leaf are illegal in a package identifier), so the package intentionally
// diverges from the path — exactly as the sibling shared surfaces do. `MatchingDeclarationName` is suppressed for
// the co-located supporting declarations.
@file:Suppress("MatchingDeclarationName", "InvalidPackageDeclaration")

package io.teslasync.android.sharedsurfaces.speed

import io.teslasync.shared.core.diagnostics.Logger
import io.teslasync.shared.core.units.UnitPref
import io.teslasync.shared.core.units.convertSpeedFromSI
import java.math.RoundingMode
import java.text.NumberFormat
import java.util.Locale

/**
 * Canonical registry metadata for the Speed surface. The diagnostics [SLUG] is emitted with the one-shot
 * `view.opened` event (P1/S11) and is the surface slug the prompt mandates (`Speed`).
 */
object SpeedRegistration {
    /** Stable surface id (also the key a host would bind the surface with). */
    const val ID: String = "speed"

    /** Diagnostics surface slug emitted with the `view.opened` event (P1/S11, prompt-mandated). */
    const val SLUG: String = "Speed"
}

/** The em-dash the web renders when no usable value is supplied (`'—'`). */
const val DASH: String = "\u2014"

/**
 * Web `fmtNumber`'s global precision default — the user's `decimal_precision`, 2 when unset. The web `Speed`
 * passes its `precision` prop straight to `fmtNumber`, which falls back to this global when the prop is absent.
 * Mirrors the sibling SpeedGearPanel's `DEFAULT_DECIMAL_PRECISION`.
 */
const val DEFAULT_DECIMAL_PRECISION: Int = 2

/** 1 mph = 0.44704 m/s exactly (web `mph * 0.44704`, NIST international yard). */
private const val MPH_TO_MPS: Double = 0.44704

/** km/h → m/s numerator / denominator (web `kmh * 1000 / 3600`). */
private const val METERS_PER_KM: Double = 1000.0
private const val SECONDS_PER_HOUR: Double = 3600.0

/** Source-unit symbol for the hover title — the caller's INPUT unit, not the display unit (web literal `mph`). */
private const val MPH_LABEL: String = "mph"

/** Source-unit symbol for the hover title — the caller's INPUT unit, not the display unit (web literal `km/h`). */
private const val KMH_LABEL: String = "km/h"

/** Fixed fraction digits for the hover title's source value (web `toFixed(1)`). */
private const val TITLE_DECIMALS: Int = 1

/**
 * The presentational inputs — the native analogue of the web `Speed` props that affect output. A caller supplies
 * the value in exactly one source unit ([mph] OR [kmh], [mph] winning when both are finite, matching the web's
 * `if (mph) … else if (kmh)`), plus an optional [precision] override. `className` is a render-layer concern and so
 * lives on the composable, not in this pure projection.
 */
data class SpeedSpec(
    val mph: Double? = null,
    val kmh: Double? = null,
    val precision: Int? = null,
)

/**
 * The fully projected, render-ready value — everything the web component derives before returning JSX.
 *
 * @property number the formatted numeric part in the user's display unit, or [DASH] when there is no value.
 * @property unitLabel the display unit symbol (web `unitPrefs.speed`), e.g. `mph` / `km/h`.
 * @property text the full visible string — `"<number> <unit>"` for a value, or just [DASH] when empty (the web
 *   em-dash span carries no unit).
 * @property title the hover-title text — the raw caller value with its SOURCE unit (web `title`), or `null` when
 *   there is no value (the web em-dash span has no `title`).
 * @property accessibleLabel the screen-reader label the composable exposes; equals [text] so a reader announces
 *   the same value the eye sees.
 * @property hasValue whether a finite caller value was supplied (false selects the em-dash empty branch).
 */
data class SpeedDisplay(
    val number: String,
    val unitLabel: String,
    val text: String,
    val title: String?,
    val accessibleLabel: String,
    val hasValue: Boolean,
)

/**
 * Pure projection from the surface's inputs to its render state — a 1:1 port of the web component's derivations:
 * the source-unit → SI conversion, the SINGLE SI → display conversion at the render site (web `toSpeedDisplay` =
 * `convertSpeedFromSI`), the `fmtNumber`-parity formatting at the resolved precision, the raw-value hover title,
 * and the em-dash empty branch. Stateless and side-effect-free so it is fully covered by the off-device unit gate;
 * the composable only collects the live unit preference and draws what these return.
 */
object SpeedProjection {
    /**
     * The caller value in SI metres-per-second — web `mph * 0.44704` (preferred) else `kmh * 1000 / 3600`, with a
     * null / non-finite value falling through exactly as the web `Number.isFinite` guards do. `null` when neither
     * source is a finite number, which selects the em-dash branch.
     */
    fun sourceMetersPerSecond(spec: SpeedSpec): Double? =
        when {
            spec.mph != null && spec.mph.isFinite() -> spec.mph * MPH_TO_MPS
            spec.kmh != null && spec.kmh.isFinite() -> spec.kmh * METERS_PER_KM / SECONDS_PER_HOUR
            else -> null
        }

    /**
     * The hover-title text — the raw caller value with its SOURCE unit (web `${mph.toFixed(1)} mph` /
     * `${kmh.toFixed(1)} km/h`), so a user can read the exact underlying figure regardless of the rounded,
     * unit-converted display. `null` when there is no finite value (the web em-dash span sets no `title`).
     */
    fun title(spec: SpeedSpec): String? =
        when {
            spec.mph != null && spec.mph.isFinite() -> "${fixedOneDecimal(spec.mph)} $MPH_LABEL"
            spec.kmh != null && spec.kmh.isFinite() -> "${fixedOneDecimal(spec.kmh)} $KMH_LABEL"
            else -> null
        }

    /**
     * The fraction digits the display number is formatted with — web `fmtNumber(value, precision)` resolution: the
     * per-call [SpeedSpec.precision] override, else the user's global precision ([UnitPref.precision]), else
     * [DEFAULT_DECIMAL_PRECISION]. Coerced non-negative so a stray negative can never reach the formatter.
     */
    fun resolvePrecision(
        spec: SpeedSpec,
        prefs: UnitPref,
    ): Int = (spec.precision ?: prefs.precision ?: DEFAULT_DECIMAL_PRECISION).coerceAtLeast(0)

    /**
     * The render-ready [SpeedDisplay] for the given [spec], display [prefs] (web `useUnits`), and [locale] (the
     * grouping/separator locale, web `fmtNumber`'s active locale). A finite caller value is converted to SI then
     * to the display unit ONCE via [convertSpeedFromSI] and formatted at the resolved precision with the unit
     * appended; an absent / non-finite value yields the em-dash branch (no unit, no title), exactly like the web.
     */
    fun display(
        spec: SpeedSpec,
        prefs: UnitPref,
        locale: Locale,
    ): SpeedDisplay {
        val unitLabel = prefs.speed.label
        val mps = sourceMetersPerSecond(spec) ?: return emptyDisplay(unitLabel)
        val number = formatNumber(convertSpeedFromSI(mps, prefs.speed), resolvePrecision(spec, prefs), locale)
        val text = "$number $unitLabel"
        return SpeedDisplay(
            number = number,
            unitLabel = unitLabel,
            text = text,
            title = title(spec),
            accessibleLabel = text,
            hasValue = true,
        )
    }

    /** The empty (no finite value) projection — the em-dash span the web renders, carrying neither unit nor title. */
    private fun emptyDisplay(unitLabel: String): SpeedDisplay =
        SpeedDisplay(
            number = DASH,
            unitLabel = unitLabel,
            text = DASH,
            title = null,
            accessibleLabel = DASH,
            hasValue = false,
        )

    /**
     * Format a number the way the web `fmtNumber(value, decimals)` does: `toLocaleString(locale, { min == max ==
     * decimals })` with grouping separators and ECMAScript `halfExpand` rounding (round half away from zero). A
     * non-finite input is coerced to 0 (web `safeNumber`) and a signed zero normalized to positive zero so a
     * `-0.0` renders "0", matching `Intl.NumberFormat`.
     */
    fun formatNumber(
        value: Double,
        decimals: Int,
        locale: Locale,
    ): String {
        val finite = if (value.isFinite()) value else 0.0
        val normalized = if (finite == 0.0) 0.0 else finite
        val formatter =
            NumberFormat.getNumberInstance(locale).apply {
                minimumFractionDigits = decimals
                maximumFractionDigits = decimals
                roundingMode = RoundingMode.HALF_UP
                isGroupingUsed = true
            }
        return formatter.format(normalized)
    }

    /**
     * The web `value.toFixed(1)` for the hover title: a fixed one-decimal, dot-separated, ungrouped string (the
     * caller's raw figure). `Locale.US` pins the dot separator the way `toFixed` always uses one, and a signed
     * zero is normalized so `-0.0` reads "0.0".
     */
    private fun fixedOneDecimal(value: Double): String {
        val normalized = if (value == 0.0) 0.0 else value
        return String.format(Locale.US, "%.${TITLE_DECIMALS}f", normalized)
    }
}

/**
 * Resolve the BCP-47 [tag] from the user's settings (web `useUnits` locale) to a [Locale], falling back to en-US
 * for a blank / absent tag — the same default the web `fmtNumber` applies when no locale is configured.
 */
internal fun resolveDisplayLocale(tag: String?): Locale = if (tag.isNullOrBlank()) Locale.US else Locale.forLanguageTag(tag)

/**
 * PII-safe diagnostics for the surface (P1/S11). Emits only the stable, dot-namespaced `view.opened` event tagged
 * with the surface [SpeedRegistration.SLUG] — never the rendered speed, its source unit, or the preference — so a
 * diagnostics line can never leak a fleet figure. Kept free of Compose so it is unit-tested with a recording
 * [Logger]; the composable calls it once per surface open.
 */
object SpeedDiagnostics {
    private const val VIEW_OPENED: String = "view.opened"
    private const val SURFACE_KEY: String = "surface"

    /** Emits the one PII-safe `view.opened` diagnostic. Call from the composable's first-composition effect. */
    fun recordViewOpened(logger: Logger) {
        logger.info(VIEW_OPENED, mapOf(SURFACE_KEY to SpeedRegistration.SLUG))
    }
}
